### DISCUSSION TITLE
Remove `isStream` boolean from stream struct

### DISCUSSION THREAD

**Discussion by andreivladbrg** [sablier-labs/flow#280] [General]
In this discussion (https://github.com/sablier-labs/flow/discussions/271), @smol-ninja correctly pointed out that we can remove the [isStream](https://github.com/sablier-labs/flow/blob/8d0e464416b2780c784e1a86042af12faac41224/src/types/DataTypes.sol#L70) boolean and rely on a `sender` check to verify whether a stream is null. I agree with his rationale, and I believe we should make this change, as there is no objective necessity for this boolean, and it would save gas in the `create` function.

---

PS: I’m opening this discussion here because it’s relevant to flow as well, and we need to make a quicker decision since the audit starts next week. This is in contrast to lockup, where we don’t have anything scheduled yet

wdyt? @sablier-labs/solidity

**Comment by PaulRBerg** [comment]
I disagree based on the rationales I provided in that discussion.

What we should do is keep `isStream` but remove the `sender == address(0)` check, just like in Lockup.

Doing so would also create symmetry between the two contracts.

> no objective necessity for this boolean

There is. As I said, a user can choose to intentionally set the sender to the zero address. **From a logical point of view, it doesn't matter that we block it in the create function**.

**Reply by andreivladbrg** [reply]
>but remove the `sender == address(0)` check, just like in Lockup.

but, we have introduced this check in lockup as well, based on your proposal here: https://github.com/sablier-labs/v2-core/discussions/38#discussioncomment-10210887

does it mean you have changed your mind? if yes i agree keeping the `isStream` boolean and remove `sender = 0` check

 > Doing so would also create symmetry between the two contracts

agree, that's an important, if we remove it here, we should remove it there

**Reply by PaulRBerg** [reply]
No, I have not changed my mind. I have always wanted symmetry — I suggested implementing the zero address check in Lockup because **I had thought it was needed in Flow**. Now that it turns out it's not needed in Flow, we should remove it in Lockup (because it's redundant).

**Reply by andreivladbrg** [reply]
> I have always wanted symmetry

agree, symetry is important

> I had thought it was needed in Flow.Now that it turns out it's not needed in Flow, we should remove it in Lockup (because it's redundant).

there is some misunderstanding from my end because:

Shub [said](https://github.com/sablier-labs/v2-core/issues/1008): "Anybody can accidentally set sender as zero address when creating the streams. Now because other functions such as cancel, renounce etc. are sender only, the stream creator would not be able to use them."

and you responded: "Agree with the general point"

do you still agree with this?

**Reply by PaulRBerg** [reply]
I don't remember the problem situation well. However, I don't think it's relevant anymore as I'm happy to keep the zero address check:

https://github.com/sablier-labs/flow/discussions/280#discussioncomment-10808820

**Reply by andreivladbrg** [reply]
great :handshake:

your reply was posted after mines, so i didn't see it prior to it, i was talking about [this comment](https://github.com/sablier-labs/flow/discussions/280#discussioncomment-10808482)

![image](https://github.com/user-attachments/assets/87c33b09-1448-400d-97af-986542c54521)

![image](https://github.com/user-attachments/assets/2993a63d-5750-445c-977a-505086d4011d)


> I don't think it's relevant anymore

you are correct, it is not, we agreed on keeping the zero check

**Reply by smol-ninja** [reply]
Great. So do we also agree to keep `isStream` field as well? (I am in favour as per https://github.com/sablier-labs/flow/discussions/280#discussioncomment-10808723).

**Reply by andreivladbrg** [reply]
yep

**Comment by smol-ninja** [comment]
> What we should do is keep isStream but remove the sender == address(0) check, just like in Lockup

We decided to implement a zero check for the sender address as per your comment [here](https://github.com/sablier-labs/v2-core/discussions/38#discussioncomment-10210887), and the [relevant PR](https://github.com/sablier-labs/v2-core/pull/1018). Therefore, regarding your symmetry argument, we have it now.

> a user can choose to intentionally set the sender to the zero address

Might be possible but I think there cannot be "any" incentive to do so. A stream with the sender as `address(0)` cannot pause it, cannot restart it, cannot refund from it. Wouldn't the same argument that applies to having a zero address check for the recipient apply here, which is, in my opinion, that accidentally setting them to the 0 address is a possible mistake?

However, logically, a stream can have 0 addresses for the sender and recipient (I am talking in theory; not talking specifically about our implementation), so defining whether a stream is valid or not based on the sender check does not seem logical to me, which, however, is not the case for `rps == 0` and `isPaused`.

**Reply by PaulRBerg** [reply]
First part — see comment above.

Second part — good point. There can, however, be incentives for setting up a stream with the address zero. Permissionless payroll, i.e., a Flow stream that is guaranteed to be paid to the recipient.

Though I think it's an esoteric use case, so we might just want to keep the check in this case, both here and in Lockup.
